qutip/tests/test_entropy.py

```diff
@@ -89,6 +90,113 @@ def test_component_selection(self):
         assert abs(qutip.entropy_mutual(dm, [0, 2], [1]) - expect) < 1e-13
 
 
+class TestRelativeEntropy:
+    def _simple_relative_entropy_implementation(
+            self, rho, sigma, log_base=np.log, tol=1e-12):
+        """ A simplified relative entropy implementation for use in
+            double-checking the optimised implementation within
+            QuTiP itself.
+        """
+        # S(rho || sigma) = sum_i(p_i log p_i) - sum_ij(p_i P_ij log q_i)
+        rvals, rvecs = sp_eigs(rho.data, rho.isherm, vecs=True)
+        svals, svecs = sp_eigs(sigma.data, sigma.isherm, vecs=True)
+        # Calculate S
+        S = 0
+        for i in range(len(rvals)):
+            if abs(rvals[i]) >= tol:
+                S += rvals[i] * log_base(rvals[i])
+            for j in range(len(svals)):
+                P_ij = (
+                    np.dot(rvecs[i], svecs[j].conjugate()) *
+                    np.dot(svecs[j], rvecs[i].conjugate())
+                )
+                if abs(svals[j]) < tol and not (
+                        abs(rvals[i]) < tol or abs(P_ij) < tol):
+                    # kernel of sigma intersects support of rho
+                    return np.inf
+                if abs(svals[j]) >= tol:
+                    S -= rvals[i] * P_ij * log_base(svals[j])
+        return np.real(S)
+
+    def test_rho_or_sigma_not_oper(self):
+        rho = qutip.bra("00")
+        sigma = qutip.bra("01")
+        with pytest.raises(TypeError) as exc:
+            qutip.entropy_relative(rho.dag(), sigma)
+        assert str(exc.value) == "Inputs must be density matrices."
+        with pytest.raises(TypeError) as exc:
+            qutip.entropy_relative(rho, sigma.dag())
+        assert str(exc.value) == "Inputs must be density matrices."
+        with pytest.raises(TypeError) as exc:
+            qutip.entropy_relative(rho, sigma)
+        assert str(exc.value) == "Inputs must be density matrices."
+
+    def test_rho_and_sigma_have_different_shape_and_dims(self):
+        rho = qutip.ket("00")
+        sigma = qutip.ket("0")
+        with pytest.raises(ValueError) as exc:
+            qutip.entropy_relative(rho, sigma)
+        assert str(exc.value) == "Inputs must have the same shape and dims."
```

Comment:
A good different dims test would be Qobjs that have compatible shapes, but incompatible dims (e.g. one in a [2, 3] space and one in a [3, 2] space) - Numpy/Scipy functions will generally just fail if shapes mismatch, but this test ensures that we're enforcing the Hilbert space constraints as well.

Reply:
Given that we only check the dims, I'm not too worried.

Reply:
Oh yeah, I'm not worried about the function as it is right now either.  I just know that matching shapes but mismatched dims is a bug we that crops up from time to time all over the place in QuTiP.  You can still make the test case in shorthand -
```python
rho = qutip.basis([2, 3], [0, 0])
sigma = qutip.basis([3, 2], [0, 0])
```

Reply:
Added!